core/backend/src/StashManager.ts

```diff
@@ -0,0 +1,458 @@
+import * as path from "path";
+import { BriefcaseManager } from "./BriefcaseManager";
+import { BriefcaseDb } from "./IModelDb";
+import { _elementWasCreated, _getHubAccess, _hubAccess, _nativeDb } from "./internal/Symbols";
+import { SQLiteDb } from "./SQLiteDb";
+import { TxnProps } from "./TxnManager";
+import { IModelHost } from "./IModelHost";
+import { BackendLoggerCategory } from "./BackendLoggerCategory";
+
+const loggerCategory = BackendLoggerCategory.StashManager;
+
+/**
+ * Custom error class for stash-related errors.
+ * @internal
+ */
+export class StashError extends Error {
+  constructor( message: string) {
+    super(message);
+    this.name = "StashError";
+    Logger.logError(loggerCategory, message);
+  }
+}
+
+/**
+ * Properties of a stash
+ * @internal
+ */
+export interface StashProps {
+  /** Unique identifier for the stash */
+  readonly id: GuidString;
+
+  /** ID of the iModel being stashed */
+  readonly iModelId: GuidString;
+
+  /** ID of the briefcase being stashed */
+  readonly briefcaseId: number;
+
+  /** ISO local Timestamp of the stash */
+  readonly timestamp: string;
+
+  /** Description of the stash */
+  readonly description: string;
+
+  /** Hash of the stash */
+  readonly hash: string;
+
+  /** Parent changeset of the stash */
+  readonly parentChangeset: ChangesetIdWithIndex;
+
+  /** ID sequences for the stash */
+  readonly idSequences: {
+    element: Id64String;
+    instance: Id64String;
+  };
+
+  /** Transaction properties for the stash */
+  readonly txns: TxnProps[];
+
+  /** Number of locks acquired by the stash */
+  readonly acquiredLocks: number;
+}
+
+/**
+ * Properties for creating a stash
+ * @internal
+ */
+export interface CreateStashProps {
+  /** Briefcase database instance */
+  readonly db: BriefcaseDb;
+  /** description of the stash */
+  readonly description: string;
+  /** discard all local changes and unless keepLocks flag is set, all locks will be released */
+  readonly discardLocalChanges?: true;
+  /** keep all locks after discarding local changes */
+  readonly retainLocks?: true;
+}
+
+/**
+ * Arguments for stash operations
+ * @internal
+ */
+export interface StashArgs {
+  readonly db: BriefcaseDb;
+  readonly stash: Id64String | StashProps;
+}
+
+/**
+ * Arguments for applying or restoring a stash.
+ *
+ * Extends {@link StashArgs} with an additional `method` property to specify the operation.
+ *
+ * @property method - Specifies the stash operation to perform.
+ *   - `"apply"`: Apply the stash without removing it.
+ *   - `"restore"`: Apply the stash and remove it from the stash list.
+ * @internal
+ */
+export interface StashApplyArgs extends StashArgs {
+  readonly method: "apply" | "restore";
+}
+
+enum LockOrigin {
+  Acquired = 0,
+  NewElement = 1,
+  Discovered = 2,
+}
+
+/**
+ * Stash manager allow stash, drop, apply and merge stashes
+ * @internal
+ */
+export class StashManager {
+
+  private static readonly STASH_ROOT_DIR_NAME: string = ".stashes";
```

Comment:
I think having a singular constant (`STASH`) with a plural value (`stashes`) is not ideal.

Reply:
fixed